# DiploDatos 2025- Mentoría " ¿Cómo clasfico a mis clientes? ML aplicado a un caso de negocio financiero"

## <font color=blue> Trabajo Practico N° 1 

##  Mentoría " Customizing offers: ¿Cómo clasifico a mis clientes? Machine Learning aplicado a un ca-so de negocio financiero"

## 1.	Familiarización con el dataset:

Para ello se deberá interpretar la documentación brindada donde se detalle el significado de cada variable, posteriormente se deberá importar los datos, analizar su estructura, tipos de campos, ver si hay datos faltantes. 

## 2.	Interpretando el dataset
¿Qué tipo de variables tenemos?, categóricas, numéricas, discretas, continuas?  

Brindar los estadisticos descriptivos, media, mediana, percentiles, de comentarios de los mismos.

¿Como es la distribucion del ingreso promedio, asimetrica, simetrica? De un comentario al respecto.

¿Hay clientes duplicados?, ¿Hay datos nulos? ¿A qué se puede deber? 

¿Hay outliers?  ¿A qué se puede deber?


## 3.	Análisis y Visualización de los datos


Para un buena predicción/segmentación es necesario realizar un buen análisis descriptivo, exploración de los datos considerando su tipo y valores definidos y tener acabado conocimiento de que representa cada uno de ellos.

Para los datos numéricos realizar una breve descripción de los mismos, analizando sus medias, percentiles, outliers. 
Mientras que para las variables categóricas realizar un análisis de transaccionalidad vrs Estado Civil, Tipo de Cuenta, UsaPlataforma, etc. Realizar un análisis de variabilidad de la transaccionalidad respecto de estas variables y otras más que puedan considerar de utilidad para el entendimiento del comportamien-to del cliente.

También para este análisis se puede considerar las variables CROSS_IND y CROSS_EMPR que nos indican cantidad de productos que ya tienen los clientes, ello nos pueda dar una pista de que tan dispuestos son de tener varios de los productos.

Algunas de las preguntas que nos podríamos realizar son las siguientes:

¿Las variables de transaccionalidad son suficientes? ¿se podrían agrupar más? ¿de qué forma? ¿Estas podrán servir para la modelización?

¿Quien tiene mas productos? recordar que para ver esto es necesario utilizar las variables Cross_Ind y  Cross_Empr, luego del analisis, realizar una conclusion, ¿están bien calculadas acorde a su definición? Estas ayudaran a la hora de realizar el modelo?

¿Es relevante el dato capita? ¿ como se comporta con respecto a a la transaccionalidad?

Al dato Edad, ¿cómo lo consideramos? Discreto o Continuo? Grafique su distribucion y de breve comentario.

¿Existe algún tipo de relación entre la edad del cliente y el uso de plataforma?

¿A qué fecha son los datos? ¿Fue correcto la selección de este periodo para el análisis? 

¿Existe algún tipo de relación entre la cantidad de productos alcanzados  y el nivel de ingresos?

¿Los datos nos permitirán realizar una clasificación de los clientes en función de sus transacciones? ¿O en función de los productos que tiene? ¿Cual sera la mas adecuada?

¿Cuál es el servicio/producto más utilizado?, recordar que esto lo podemos obtener de las variables Tiene_i para los productos y de las variables indicadas en la sección Información sobre Cantidad de Transaccionalidad Realizada.

¿Qué tipos de clientes, con respecto a la segmentación dada (segmento_sueldo y segmen-to_bancas, subgrupo) utilizan más nuestros productos?

¿La edad, estado civil, sexo son datos relevantes? ¿ Se puede ver alguna correlacion con los productos/servicios?

¿El ingreso promedio de los últimos 6 meses, se puede considerar como relevante? Nos podrá ayudar en nuestra segmentación? 

¿Qué producto/servicio se ha utilizado más? Hay alguna relación con el periodo (¿mes y año utilizados?)

¿Qué clientes han realizado más depósitos?

¿Sobre las transferencia, que tipo de trasferencia se ha realizado más? ¿Enviada o Recibida? Y en cuento a los montos cual ha sido el mayor? ¿Cuál servicio tuvo el mayor monto, Modo, Home-Banking, ATM, resto? 

¿Qué tipos de clientes realizan operatoria de inversión en títulos/FCI?

Realizar un ranking de los 10 clientes con mas productos y analizar su transaccionalidad.

¿Algún cliente ha dado indicio de dar de baja nuestros servicios, es decir de cerrar su cuenta? 

¿Como es la distribución de los datos cross_ind, cross_empr recalculados con respecto a las variables de transaccionalidad, por ejemplo las de compras, las de transferencias observamos que muchos de nuestros clientes que tienen más de un producto utiliza estos servicios?

Como es el comportamiento de aquellos clientes que han realizado Plazo Fijo, con respecto al resto de la transaccionalidad, ¿realizan también inversión en Titulos/FCI? Realizar análisis simila-res para el resto de la operatividad.

Basándose en los análisis anteriores, realice una matriz de correlacion entre las variables.

Puede ya definir cuales serán las variables mas interesantes que contribuirán al modelo.


# Importar librerías necesarias

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statistics as stat
import scipy.stats as st
from statistics import stdev
from scipy.stats import t
import datetime
#import missingno as msno
import seaborn

seaborn.set_context('talk')
sns.set_context('talk')

# Seteo de displays
pd.set_option('display.max_columns',10000)
pd.set_option('display.max_rows',10000)
pd.set_option('display.float_format', lambda x: '%.6f' % x)
pd.options.display.max_columns = None # para que no recorte las columnas

# Leer el dataset

In [2]:
# Importamos el datset que definimmos

df = pd.read_csv ("muestra.csv", sep=';', encoding='latin-1')

In [3]:
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2831 entries, 0 to 2830
Columns: 126 entries, id to Unnamed: 125
dtypes: float64(108), int64(6), object(12)
memory usage: 2.7+ MB
None


Como se puede ver hay 125 variables de las cuales 88 son numericas con decimales, 24 numericas enteras y 13 variables categoricas.

In [4]:
df.head()

id            tipo_cuenta_ca_pesos tipo_cuenta_ca_dolar  \
0   1  Caja Ahorro Jubilado Provincia                  NaN   
1   2   Cuenta de la Seguridad Social                  NaN   
2   3   Cuenta de la Seguridad Social                  NaN   
3   4   Cuenta de la Seguridad Social                  NaN   
4   5   Cuenta de la Seguridad Social                  NaN   

  tipo_cuenta_cc_pesos sexocodi   estcivil  edad  plataforma_emp_um  \
0                  NaN        M  Soltero/a    69                  0   
1                  NaN        M    Viudo/a    69                  0   
2                  NaN        M  Soltero/a    69                  0   
3                  NaN        M  Soltero/a    69                  0   
4                  NaN        M  Soltero/a    69                  0   

  usa_plataforma_u3m  ingr_prom_u6m  capita grupo_cliente         subgrupo  \
0                 NO         366085  PROPIA            I4           Sueldo   
1                 NO         468618  PROPIA            I4           Sueldo   
2                 NO         298270  PROPIA            I4           Sueldo   
3                 NO         367273  PROPIA            I4           Sueldo   
4                 NO          13295      NO            I8  Publico General   

  relacion       fum       segmento_sueldos        segmento_bancas  es_capita  \
0      TIT  20241129    JUBILADOS PROVINCIA    JUBILADOS PROVINCIA          1   
1      TIT  20241127        JUBILADOS ANSES        JUBILADOS ANSES          1   
2      TIT  20241130        JUBILADOS ANSES        JUBILADOS ANSES          1   
3      TIT  20241129  SUELDO ADM PUBL RESTO  SUELDO ADM PUBL RESTO          1   
4      TIT  20240503        JUBILADOS ANSES                  OTROS          0   

   tiene_i_cbesa  tiene_i_cs  tiene_i_pr_hip  tiene_i_pr_pers  tiene_i_prep  \
0       1.000000    0.000000        0.000000         0.000000      0.000000   
1       0.000000    0.000000        0.000000         0.000000      0.000000   
2       0.000000    0.000000        0.000000         1.000000      0.000000   
3       1.000000    0.000000        0.000000         0.000000      0.000000   
4       0.000000    0.000000        0.000000         0.000000      0.000000   

   tiene_i_se_patr  tiene_i_visa  tiene_i_tj_soc  tiene_e_ac_cc  tiene_e_fact  \
0         0.000000      0.000000        0.000000       0.000000      0.000000   
1         0.000000      0.000000        0.000000       0.000000      0.000000   
2         0.000000      0.000000        0.000000       0.000000      0.000000   
3         0.000000      0.000000        0.000000       0.000000      0.000000   
4         0.000000      0.000000        0.000000       0.000000      0.000000   

   saldo_computo_sist_fci  tiene_e_liq_com  tiene_i_seg_vida  tiene_e_tc_agro  \
0                0.000000         0.000000          0.000000         0.000000   
1                0.000000         0.000000          0.000000         0.000000   
2                0.000000         0.000000          0.000000         0.000000   
3                0.000000         0.000000          0.000000         0.000000   
4                0.000000         0.000000          0.000000         0.000000   

   tiene_e_tc_corpo  tiene_e_tc_flota  saldo_computo_sist_2  \
0          0.000000          0.000000              0.000000   
1          0.000000          0.000000              0.000000   
2          0.000000          0.000000              0.000000   
3          0.000000          0.000000              0.000000   
4          0.000000          0.000000              0.000000   

   saldo_computo_sist_ca_p  saldo_computo_cj  cross_empr  cross_ind  \
0             64636.460000          0.000000    0.000000   3.000000   
1               624.560000          0.000000    0.000000   2.000000   
2            918728.770000          0.000000    0.000000   3.000000   
3            721193.640000          0.000000    0.000000   3.000000   
4               203.300000          0.000000    0.000000   1.000

In [5]:
df.describe()

id        edad  plataforma_emp_um   ingr_prom_u6m  \
count 2831.000000 2831.000000        2831.000000     2831.000000   
mean  1416.000000   50.202402           0.002826   264368.066761   
std    817.383631   20.263485           0.053093   921126.722738   
min      1.000000    6.000000           0.000000        0.000000   
25%    708.500000   35.000000           0.000000     5000.000000   
50%   1416.000000   49.000000           0.000000    16176.000000   
75%   2123.500000   63.000000           0.000000   308341.500000   
max   2831.000000  123.000000           1.000000 34590551.000000   

                  fum   es_capita  tiene_i_cbesa  tiene_i_cs  tiene_i_pr_hip  \
count     2831.000000 2831.000000    2831.000000 2831.000000     2831.000000   
mean  18366705.459202    0.289650       0.296362    0.007065        0.006711   
std    5854148.238105    0.453680       0.456733    0.083769        0.081662   
min          0.000000    0.000000       0.000000    0.000000        0.000000   
25%   20220623.000000    0.000000       0.000000    0.000000        0.000000   
50%   20241119.000000    0.000000       0.000000    0.000000        0.000000   
75%   20241129.000000    1.000000       1.000000    0.000000        0.000000   
max   20241130.000000    1.000000       1.000000    1.000000        1.000000   

       tiene_i_pr_pers  tiene_i_prep  tiene_i_se_patr  tiene_i_visa  \
count      2831.000000   2831.000000      2831.000000   2831.000000   
mean          0.085482      0.024373         0.020134      0.055104   
std           0.279647      0.154232         0.140484      0.228224   
min           0.000000      0.000000         0.000000      0.000000   
25%           0.000000      0.000000         0.000000      0.000000   
50%           0.000000      0.000000         0.000000      0.000000   
75%           0.000000      0.000000         0.000000      0.000000   
max           1.000000      1.000000         1.000000      1.000000   

       tiene_i_tj_soc  tiene_e_ac_cc  tiene_e_fact  saldo_computo_sist_fci  \
count     2831.000000    2831.000000   2831.000000             2831.000000   
mean         0.026139       0.002473      0.000353            29739.180148   
std          0.159577       0.049673      0.018794           908874.783920   
min          0.000000       0.000000      0.000000                0.000000   
25%          0.000000       0.000000      0.000000                0.000000   
50%          0.000000       0.000000      0.000000                0.000000   
75%          0.000000       0.000000      0.000000                0.000000   
max          1.000000       1.000000      1.000000         45066721.000000   

       tiene_e_liq_com  tiene_i_seg_vida  tiene_e_tc_agro  tiene_e_tc_corpo  \
count      2831.000000       2831.000000      2831.000000       2831.000000   
mean          0.004592          0.063229         0.000706          0.001060   
std           0.067621          0.243417         0.026575          0.032541   
min           0.000000          0.000000         0.000000          0.000000   
25%           0.000000          0.000000         0.000000          0.000000   
50%           0.000000          0.000000         0.000000          0.000000   
75%           0.000000          0.000000         0.000000          0.000000   
max           1.000000          1.000000         1.000000          1.000000   

       tiene_e_tc_flota  saldo_computo_sist_2  saldo_computo_sist_ca_p  \
count       2831.000000           2831.000000              2831.000000   
mean           0.000000         212526.938107            166509.029657   
std            0.000000        2640445.679965            899028.942667   
min            0.000000              0.000000                 0.000000   
25%            0.000000              0.000000                 1.825000   
50%            0.000000              0.000000              1012.730000   
75%            0.000000              0.000000             57751.015000   
max            0.000000      10144